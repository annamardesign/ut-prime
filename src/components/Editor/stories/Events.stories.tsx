import React from 'react';
import type { Story } from '@storybook/react';
import type { Props } from '../Editor.types';
import Editor from '..';
import useToast from '../../hooks/useToast';

import {input, dropdowns} from '../../test/input';

export {default} from '../Editor.stories';

const sticky = {sticky: false};

const Template: Story<Props> = ({methods, ...args}) => {
    const {toast, submit} = useToast(sticky);
    return <>
        <Editor
            id={1}
            object='input'
            onGet={() => Promise.resolve({input: {}})}
            onDropdown={() => Promise.resolve(dropdowns)}
            onAdd={submit}
            onEdit={submit}
            onFieldChange='handleFieldChange'
            methods={{
                async handleFieldChange({field, value, event}: {field: unknown, value, event: Event}) {
                    submit({field, value});
                    // throw new Error('test error');
                    // return false;
                },
                async handleAutocomplete() {
                    return [{
                        value: 1, label: 'value 1'
                    }, {
                        value: 2, label: 'value 2'
                    }, {
                        value: 3, label: 'value 3'
                    }];
                },
                async 'portal.customization.get'() {
                    return {};
                },
                ...methods
            }}
            {...args}
        />
        {toast}
    </>;
};

declare type StoryTemplate = Story<Partial<Props>> & {
    play: (context: {canvasElement: HTMLElement}) => Promise<void>
}

export const Events: StoryTemplate = Template.bind({});
Events.args = {
    ...input,
    layouts: {edit: ['left', 'center', 'right']}
};

export const EventsFormAPI: StoryTemplate = Template.bind({});
EventsFormAPI.args = {
    schema: {
        properties: {
            a: {
                type: 'integer',
                widget: {onChange: 'handleA'}
            },
            b: {
                type: 'integer',
                widget: {onChange: 'handleB'}
            },
            sum: {
                type: 'number',
                readOnly: true
            },
            discount: {
                type: 'number',
                widget: {onChange: 'handleDiscount'}
            },
            bitmask: {
                widget: {onChange: 'handleBitmask'}
            }
        }
    },
    cards: {
        edit: {
            className: 'lg:col-3',
            widgets: [
                'a',
                'b',
                'sum',
                'discount',
                'bitmask'
            ]
        }
    },
    onGet: () => Promise.resolve({a: 0, b: 0, sum: 0}),
    methods: {
        async handleA({form, value}) {
            form.setValue('sum', Number(form.getValues('b')) + Number(value));
        },
        async handleB({form, value}) {
            form.setValue('sum', Number(form.getValues('a')) + Number(value));
        },
        async handleDiscount({value, form}) {
            if (Number(value) < 0) throw new Error('Discount must be > 0'); // throwing means abort the edit and show error
            if (Number(value) >= 0) form.clearErrors('discount');
        },
        async handleBitmask({value}) {
            if (/[^01]/.test(value)) return false; // returning false means abort the edit silently
        }
    }
};
