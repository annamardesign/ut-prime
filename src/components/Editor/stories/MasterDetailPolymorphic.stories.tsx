import { userEvent, within } from '@storybook/testing-library';
import React from 'react';
import Editor from '..';
export {default} from '../Editor.stories';

const common = {
    className: 'xl:col-3',
    label: 'Shape properties',
    watch: '$.selected.shape'
};

export const MasterDetailPolymorphic = () =>
    <Editor
        id={1}
        onGet={() => Promise.resolve({
            shape: [
                {name: 'Big circle', type: 'circle', radius: 100},
                {name: 'Small ellipse', type: 'ellipse', radius: 5, secondRadius: 4},
                {name: 'Big square', type: 'square', side1: 100},
                {name: 'Medium triangle', type: 'triangle', side1: 30, side2: 40, side3: 50},
                {name: 'Small rectangle', type: 'rectangle', side1: 2, side2: 4}
            ]
        })}
        onDropdown={() => Promise.resolve({})}
        schema={{
            type: 'object',
            properties: {
                shape: {
                    type: 'array',
                    items: {
                        type: 'object',
                        properties: {
                            name: {},
                            type: {enum: ['circle', 'square', 'triangle', 'rectangle', 'ellipse']},
                            radius: {title: 'Radius', type: 'number'},
                            secondRadius: {title: 'Second radius', type: 'number'},
                            side1: {title: 'Side Length', type: 'number'},
                            side2: {title: 'Second Side Length', type: 'number'},
                            side3: {title: 'Third Side Length', type: 'number'},
                            tags: {}
                        }
                    },
                    title: '',
                    widget: {
                        type: 'table',
                        selectionMode: 'single',
                        actions: {
                            allowEdit: false
                        }
                    }
                }
            }
        }}
        cards={{
            shapes: {
                label: 'Shapes',
                className: 'xl:col-2',
                widgets: [{
                    name: 'shape',
                    widgets: ['name']
                }]
            },
            shapeCircle: {
                ...common,
                match: {type: 'circle'},
                widgets: [
                    '$.edit.shape.radius'
                ]
            },
            shapeEllipse: {
                ...common,
                match: {type: 'ellipse'},
                widgets: [
                    '$.edit.shape.radius',
                    '$.edit.shape.secondRadius'
                ]
            },
            shapeSquare: {
                ...common,
                match: {type: 'square'},
                widgets: [
                    '$.edit.shape.side1'
                ]
            },
            shapeRectangle: {
                ...common,
                match: {type: 'rectangle'},
                widgets: [
                    '$.edit.shape.side1',
                    '$.edit.shape.side2'
                ]
            },
            shapeTriangle: {
                ...common,
                match: {type: 'triangle'},
                widgets: [
                    '$.edit.shape.side1',
                    '$.edit.shape.side2',
                    '$.edit.shape.side3'
                ]
            }
        }}
        layouts={{
            edit: ['shapes', ['shapeCircle', 'shapeSquare', 'shapeTriangle', 'shapeRectangle', 'shapeEllipse']]
        }}
    />;

MasterDetailPolymorphic.play = async({canvasElement}) => {
    const canvas = within(canvasElement);
    userEvent.click(canvas.getByText('Small ellipse'));
};
