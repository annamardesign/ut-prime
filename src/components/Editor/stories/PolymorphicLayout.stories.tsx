import React from 'react';
import Editor from '..';
export {default} from '../Editor.stories';

export const PolymorphicLayout = () => {
    const props = {
        typeField: 'type',
        keyField: 'id',
        onDropdown: names => Promise.resolve({}),
        schema: {
            properties: {
                name: {},
                role: {},
                registrationNumber: {}
            }
        },
        cards: {
            editPerson: {
                label: 'Person',
                className: 'col-12',
                widgets: ['name', 'role']
            },
            editOrganization: {
                label: 'Organization',
                className: 'col-12',
                widgets: ['name', 'registrationNumber']
            }
        }
    };

    return (<div className='flex flex-row'>
        <div className='flex-col flex-grow-1'>
            <Editor
                id={1}
                onGet={() => Promise.resolve({
                    id: 1,
                    type: 'person',
                    name: 'John Doe',
                    role: 'Director'
                })}
                {...props}
            />
        </div>
        <div className='flex-col flex-grow-1'>
            <Editor
                id={2}
                onGet={() => Promise.resolve({
                    id: 2,
                    type: 'organization',
                    name: 'Acme Inc.',
                    registrationNumber: '111-111-111'
                })}
                {...props}
            />
        </div>
    </div>);
};
