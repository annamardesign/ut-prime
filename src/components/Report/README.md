# Report

## How to use

```jsx
import Report from 'ut-front-devextreme/core/Report';

<Report className='name' />;
```

## Props

- **className** - class applied to root `Report` html element.

| propName  | propType | defaultValue | isRequired |
| --------- | -------- | ------------ | ---------- |
| className | string   |              | no         |
