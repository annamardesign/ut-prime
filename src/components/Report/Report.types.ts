import {WithStyles, withStyles, createStyles} from '@material-ui/core/styles';
import React from 'react';
import type { Schema } from 'joi';

import type {Properties, Dropdowns} from '../types';

export interface Props {
    properties: Properties,
    validation?: Schema,
    params: string[],
    columns: string[],
    resultSet?: string,
    onDropdown: (params: string[]) => Promise<Dropdowns>,
    fetch: (params: {}) => Promise<{
        pagination?: {
            recordsTotal: number
        },
        [data: string]: any
    }>;
}

const styles = createStyles({
    component: {
    }
});

export const Styled = withStyles(styles);
export type StyledType = React.FC<Props & WithStyles<typeof styles>>
