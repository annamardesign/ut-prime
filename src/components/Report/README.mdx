# Report
import { Canvas, Story, ArgsTable } from "@storybook/addon-docs";

## How to use

```jsx
import Report from 'ut-prime/core/Report';

<Report className='name' />;
```

<Canvas>
    <Story id="report--basic" />
</Canvas>

## Properties

<ArgsTable story="."/>
