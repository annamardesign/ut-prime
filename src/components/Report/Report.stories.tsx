import React from 'react';
import { withReadme } from 'storybook-readme';
import Joi from 'joi';

// @ts-ignore: md file and not a module
import README from './README.md';
import Report from './index';
import { Properties } from '../types';

export default {
    title: 'Report',
    component: Report,
    decorators: [withReadme(README)],
    args: {
        state: {}
    }
};

const columns: Properties = {
    name: {
        title: 'Name'
    },
    startDate: {
        title: 'Start Date',
        validation: Joi.date(),
        widget: {
            type: 'date'
        }
    },
    endDate: {
        title: 'End Date',
        validation: Joi.date(),
        widget: {
            type: 'date'
        }
    }
};

export const Basic = () =>
    <Report
        schema={{
            properties: {
                params: {
                    properties: columns
                },
                result: {
                    properties: columns
                }
            }
        }}
        params={['name', 'startDate', 'endDate']}
        columns={['name']}
        onDropdown={names => Promise.resolve({})}
        fetch={params => Promise.resolve({
            result: [
                {name: 'row 1'},
                {name: 'row 2'}
            ]
        })}
    />;
