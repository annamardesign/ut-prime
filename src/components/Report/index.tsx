import React from 'react';
import lodashGet from 'lodash.get';

import { Styled, StyledType } from './Report.types';

import { Button } from '../prime';
import Form from '../Form';
import getValidation from '../Form/schema';
import Explorer from '../Explorer';
import useLoad from '../hooks/useLoad';
import type {DataTableProps} from '../prime';

const flexGrow = {flexGrow: 1};
const paramsLayout = ['params'];
const table: DataTableProps = {paginatorPosition: 'top'};

const Report: StyledType = ({
    classes,
    schema,
    params = [],
    columns = [],
    init = {},
    fetch,
    onDropdown,
    resultSet = 'result'
}) => {
    const [trigger, setTrigger] = React.useState();
    const [dropdowns, setDropdown] = React.useState({});

    const [validation, dropdownNames] = React.useMemo(() => {
        const validation = getValidation(schema?.properties?.params, params);
        const dropdownNames = params
            .map(name => {
                const property = lodashGet(schema?.properties?.params?.properties, name?.replace(/\./g, '.properties.'));
                return [
                    property?.widget?.dropdown,
                    property?.widget?.pivot?.dropdown
                ];
            })
            .flat()
            .filter(Boolean);
        return [validation, dropdownNames];
    }, [schema, params]);

    const cards = React.useMemo(() => ({params: {widgets: params, className: 'col-12', flex: 'col-12 md:col-4 xl:col-3 mb-0'}}), [params]);

    useLoad(async() => {
        setDropdown(await onDropdown(dropdownNames));
    });
    const [filter, setFilter] = React.useState<[{}] | [{}, {files: []}]>([init]);
    const explorerFilter = React.useMemo(() => [
        {
            [resultSet]: filter[0]
        },
        {
            ...filter[1],
            files: filter?.[1]?.files?.map(name => `${resultSet}.${name}`)
        }
    ], [filter, resultSet]);
    return (
        <>
            <div className={`flex align-items-center ${classes.report}`}>
                <Form
                    style={flexGrow}
                    schema={schema?.properties?.params}
                    validation={validation}
                    cards={cards}
                    layout={paramsLayout}
                    onSubmit={setFilter}
                    value={filter[0]}
                    dropdowns={dropdowns}
                    setTrigger={setTrigger}
                    triggerNotDirty
                    autoSubmit
                />
                <Button className='col-1' icon='pi pi-search' onClick={trigger} disabled={!trigger}/>
            </div>
            <Explorer
                fetch={filter.length > 1 && fetch}
                schema={schema?.properties?.result}
                columns={columns}
                resultSet={resultSet}
                details={false}
                filter={explorerFilter[0]}
                index={explorerFilter[1]}
                showFilter={false}
                pageSize={20}
                table={table}
            />
        </>
    );
};

export default Styled(Report);
