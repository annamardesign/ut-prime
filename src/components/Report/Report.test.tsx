import React from 'react';

import { render } from '../test';
import { Basic } from './Report.stories';

describe('<Report />', () => {
    it('render equals snapshot', () => {
        const { getByTestId } = render(<Basic />);
        expect(getByTestId('ut-front-test')).toMatchSnapshot();
    });
});
