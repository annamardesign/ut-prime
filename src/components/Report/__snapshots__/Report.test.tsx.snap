// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Report /> render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  <div
    class="flex align-items-center Report-report-2"
  >
    <div
      class="p-toast p-component p-toast-top-right"
    >
      <div />
    </div>
    <form
      class="grid col align-self-start Form-component-3"
      style="flex-grow: 1;"
    >
      <div
        class="col-12 col-12"
      >
        <div>
          <div
            class="p-card p-component card mb-3"
            id="params"
          >
            <div
              class="p-card-body"
            >
              <div
                class="p-card-content"
              >
                <div
                  class="flex flex-wrap"
                >
                  <div
                    class="field grid col-12 md:col-4 xl:col-3 mb-0"
                  >
                    <label
                      class="col-12 md:col-4"
                      for="name"
                    >
                      Name
                    </label>
                    <div
                      class="col-12 md:col-8"
                    >
                      <input
                        class="p-inputtext p-component p-filled w-full"
                        id="name"
                        name="name"
                        value="row 1"
                      />
                    </div>
                  </div>
                  <div
                    class="field grid col-12 md:col-4 xl:col-3 mb-0"
                  >
                    <label
                      class="col-12 md:col-4"
                      for="startDate"
                    >
                      Start Date
                    </label>
                    <div
                      class="col-12 md:col-8"
                    >
                      <span
                        class="p-calendar p-component p-inputwrapper w-full p-calendar-w-btn p-calendar-w-btn-right"
                        id="startDate"
                      >
                        <input
                          autocomplete="off"
                          class="p-inputtext p-component"
                          id="startDate"
                          inputmode="none"
                          name="startDate"
                          type="text"
                        />
                        <button
                          class="p-button p-component p-datepicker-trigger p-button-icon-only"
                          tabindex="-1"
                          type="button"
                        >
                          <span
                            class="p-button-icon p-c pi pi-calendar"
                          />
                          <span
                            class="p-button-label p-c"
                          >
                             
                          </span>
                        </button>
                      </span>
                    </div>
                  </div>
                  <div
                    class="field grid col-12 md:col-4 xl:col-3 mb-0"
                  >
                    <label
                      class="col-12 md:col-4"
                      for="endDate"
                    >
                      End Date
                    </label>
                    <div
                      class="col-12 md:col-8"
                    >
                      <span
                        class="p-calendar p-component p-inputwrapper w-full p-calendar-w-btn p-calendar-w-btn-right"
                        id="endDate"
                      >
                        <input
                          autocomplete="off"
                          class="p-inputtext p-component"
                          id="endDate"
                          inputmode="none"
                          name="endDate"
                          type="text"
                        />
                        <button
                          class="p-button p-component p-datepicker-trigger p-button-icon-only"
                          tabindex="-1"
                          type="button"
                        >
                          <span
                            class="p-button-icon p-c pi pi-calendar"
                          />
                          <span
                            class="p-button-label p-c"
                          >
                             
                          </span>
                        </button>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
    <button
      class="p-button p-component col-1 p-button-icon-only"
    >
      <span
        class="p-button-icon p-c pi pi-search"
      />
      <span
        class="p-button-label p-c"
      >
         
      </span>
    </button>
  </div>
  <div
    class="flex flex-column h-full Explorer-component-4"
  >
    <div
      class="p-datatable p-component p-datatable-auto-layout p-datatable-responsive-stack"
      data-scrollselectors=".p-datatable-wrapper"
      pr_id_1=""
    >
      <div
        class="p-paginator p-component p-paginator-top"
      >
        <button
          class="p-paginator-first p-paginator-element p-link p-disabled"
          disabled=""
          type="button"
        >
          <span
            class="p-paginator-icon pi pi-angle-double-left"
          />
        </button>
        <button
          class="p-paginator-prev p-paginator-element p-link p-disabled"
          disabled=""
          type="button"
        >
          <span
            class="p-paginator-icon pi pi-angle-left"
          />
        </button>
        <span
          class="p-paginator-pages"
        />
        <button
          class="p-paginator-next p-paginator-element p-link p-disabled"
          disabled=""
          type="button"
        >
          <span
            class="p-paginator-icon pi pi-angle-right"
          />
        </button>
        <button
          class="p-paginator-last p-paginator-element p-link p-disabled"
          disabled=""
          type="button"
        >
          <span
            class="p-paginator-icon pi pi-angle-double-right"
          />
        </button>
      </div>
      <div
        class="p-datatable-wrapper"
      >
        <table
          class="p-datatable-table"
          role="table"
        >
          <thead
            class="p-datatable-thead"
          >
            <tr
              role="row"
            >
              <th
                class=""
                role="columnheader"
              >
                <div
                  class="p-column-header-content"
                >
                  <span
                    class="p-column-title"
                  >
                    <span
                      data-testid="result.nameTitle"
                    >
                      Name
                    </span>
                  </span>
                </div>
              </th>
            </tr>
            <tr
              role="row"
            >
              <th
                class="p-filter-column"
              />
            </tr>
          </thead>
          <tbody
            class="p-datatable-tbody"
          >
            <tr
              class="p-datatable-emptymessage"
              role="row"
            >
              <td
                colspan="1"
                role="cell"
              >
                No available options
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
